lib/galaxy/managers/datasets.py

```
@@ -684,6 +686,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
         # ...?
         return unwrapped_val
 
+    def deserialize_datatype(self, item, key, val, **context):
+        if item.datatype.allow_datatype_change and context['trans'].app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            if managers.hdas.HDAManager(self.app).ok_to_edit_metadata(item.dataset_id):
+                context['trans'].app.datatypes_registry.change_datatype(item, val)
+                context['trans'].sa_session.flush()
+                context['trans'].app.datatypes_registry.set_external_metadata_tool.tool_action.execute(context['trans'].app.datatypes_registry.set_external_metadata_tool, context['trans'], incoming={'input1': item}, overwrite=False)  # overwrite is False as per existing behavior
+                return item.datatype
+        context['trans'].response.status = 500
+        context['trans'].error_message = 'Datatype could not be updated'
+        return False
```

Should raise an exception instead.